### PR TITLE
Create image tutorial

### PR DESCRIPTION
**Pull request by vincentschen**
Write LFs over image primitives in this tutorial.

### CODE REVIEW

**Comment by ajratner** [comment]
@vincentschen looks like checks are failing... tag me when fixed?

**Comment by vincentschen** [comment]
@ajratner tests passing!

File: tutorials/images/Images_Tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    "\n",
+    "loader.show_examples(annotated=False, label=1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Generate Primitives\n",
+    "Since we cannot check relations in the images directly by writing labeling functions over the primitives, we first extract a set of  \"primitives\" from the images that are easily interpretable.\n",
```
Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: This sentence is confusing- 'we can't check relations by writing LFs over primitives, so we load primitives?'  Also I would avoid saying that we "cannot" write LFs over images, since this might confuse people (since we've said LFs are just generic functions). I would instead say that to make it practical, need primitives to write over / use similar language as in Coral @paroma ?

File: tutorials/images/Images_Tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+    "\n",
+    "In this case, we use information about the objects in the data such as their labels, their positions and their size.  We show how we develop simple primitives that are based on the labels of the different objects and include the more complex primitives in `primitive_helpers.py`."
```
Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: Also: make it clear that these primitives can be generated by generic off the shelf tools that are more broadly applicable then our specific problem- and that they can be noisy!

File: tutorials/images/Images_Tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    "                return 1\n",
+    "    else:\n",
+    "        return -1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Assign Heuristic Functions**\n",
```
Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: Say `Labeling functions / LFs` to be consistent with earlier in the tutorial & rest of Snorkel code

Comment (ajratner) [reply tutorials/images/Images_Tutorial.ipynb]: (comment applies throughout!)

File: tutorials/images/Images_Tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    "print ('Coverage of Majority Vote on Train Set: ', np.sum(np.sign(np.sum(np.abs(L.T),1)) != 0)/float(loader.train_num))\n",
+    "print ('Accuracy of Majority Vote on Train Set: ', np.sum(mv_labels == loader.train_ground)/float(loader.train_num))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Generative Model**\n",
+    "We assume that the heuristic functions are conditionally independent of the true label and train the generative model using the labels assigned by the heuristic functions"
```
Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: Conditionally independent *given* the true label

Comment (ajratner) [reply tutorials/images/Images_Tutorial.ipynb]: + Here might be a good place to put a pointer in to Coral @paroma e.g. "but, to see more advanced stuff we can do, see..."

File: tutorials/images/Images_Tutorial.ipynb (hunk 1 above)
Comment (paroma) [reply tutorials/images/Images_Tutorial.ipynb]: +1 maybe as "we write LFs over primitives of images instead of the raw pixel values because..."

File: tutorials/images/Images_Tutorial.ipynb (hunk 2 above)
Comment (paroma) [reply tutorials/images/Images_Tutorial.ipynb]: +1. also would be nice to mention we define a limited set of primitives here but technically we could use other information about objects in the image etc.

File: tutorials/images/Images_Tutorial.ipynb [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+   "source": [
+    "import sys\n",
+    "!{sys.executable} -m pip install scikit-image"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load and Visualize Dataset\n",
```
[the remark this answers is not in the document]
Comment (vincentschen) [reply tutorials/images/Images_Tutorial.ipynb]: This is the same as your implementation in the Coral notebook -- we load in a bunch of `.npy` files that are saved in the `tutorial_data` folder!

**Comment by vincentschen** [comment]
@paroma @ajratner A few of the comments referred to using built-in Snorkel paradigms for applying loading data, applying label matrices, printing stats, etc. 

Is this the best practice when using Snorkel? Or is it okay to diverge and show examples of manually-constructed primitives/LFs as we do in the notebook?

File: tutorials/images/Images_Tutorial.ipynb [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "labels = 2 * (train_marginals > 0.9) - 1\n",
+    "print ('Coverage of Generative Model on Train Set:', np.sum(train_marginals != 0.5)/float(len(train_marginals)))\n",
+    "print ('Accuracy of Generative Model on Train Set:', np.mean(labels == loader.train_ground))"
+   ]
+  },
```
[the remark this answers is not in the document]
Comment (jason-fries) [reply tutorials/images/Images_Tutorial.ipynb]: @vincentschen you can include a pointer to the noise-aware CNN from the MRI project (the non-video DenseNet)

[the remark this answers is not in the document]
Comment (vincentschen) [reply tutorials/images/Images_Tutorial.ipynb]: yes -- @jason-fries do you want to point to one of our papers? if so which one?

**Comment by vincentschen** [comment]
@ajratner -- @paroma's suggested Snorkel data structures in some cases!
- https://github.com/HazyResearch/snorkel/pull/969#discussion_r199338368
- https://github.com/HazyResearch/snorkel/pull/969#discussion_r199338419
- https://github.com/HazyResearch/snorkel/pull/969#discussion_r199338465
- https://github.com/HazyResearch/snorkel/pull/969#discussion_r199338484

File: tutorials/images/Images_Tutorial.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (jason-fries) [reply tutorials/images/Images_Tutorial.ipynb]: @vincentschen I'd go with @ajratner 's suggestion and just toss in some really generic CNN model (AlexNet or VGG16 from the MRI project). Those models assume no Snorkel infrastructure, just probabilistic labels so they appealing from that angle.

File: tutorials/images/Images_Tutorial.ipynb [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+    "    L[2,i] = L_names[2](P.primitive_mtx[i,0],P.primitive_mtx[i,3], P.primitive_mtx[i,4])\n",
+    "    L[3,i] = L_names[3](P.primitive_mtx[i,0],P.primitive_mtx[i,3], P.primitive_mtx[i,5])\n",
+    "    L[4,i] = L_names[4](P.primitive_mtx[i,0],P.primitive_mtx[i,3], P.primitive_mtx[i,6])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Calculate and Show Accuracy and Coverage of Heuristic Functions**"
```
[the remark this answers is not in the document]
Comment (ajratner) [reply tutorials/images/Images_Tutorial.ipynb]: 👍 Yup!  agree here

File: tutorials/images/Images_Tutorial.ipynb [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "learned_table = gen_model.learned_lf_stats()\n",
```
[the remark this answers is not in the document]
Comment (ajratner) [reply tutorials/images/Images_Tutorial.ipynb]: @paroma I think this is the helper (`learned_lf_stats`)?  Or did you have something else in mind?

File: tutorials/images/Images_Tutorial.ipynb [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "L_names = [LF_street,LF_vehicles,LF_distance,LF_size,LF_number]\n",
+    "\n",
+    "L = np.zeros((len(L_names),loader.train_num)).astype(int)\n",
+    "for i in range(loader.train_num):\n",
```
[the remark this answers is not in the document]
Comment (ajratner) [reply tutorials/images/Images_Tutorial.ipynb]: I think that showing a way to create a label matrix outside of the `LabelAnnotator` is good- just (a) get rid of the `LabelAnnotator` completely then bc no use for it, just complicates things this way, and (b) say eg 'Usually we use the `LabelAnnotator`, but here we show how to manually construct a label matrix to feed into the generative model as an alternative that is useful in some cases.'

File: tutorials/images/Images_Tutorial.ipynb [hunk 10]
```diff
@@ -0,0 +1,779 @@
...
...XEfkrgPFJncUrjNAwCj4IBTO40lbqMSHmYrsGojHJ9mDjpU/OK9JVBk2KsoGs1vjAIK3BOkFqLZwXOzw0cFIIs1Rgc1jqEJ/EvJcKnlxy3JsabLM7NgnVMjIyyenoOVSiQJgm+UyRZShDmAlCtNXHSweoMoXyEdShPYRKDthqlJMovMlQpstJuI1BYa5C+xPVTQl/klFaliGxEQkaKJXAu74YJh8XLM/p8D4iR1SpKeEijUCi08MiyJYxXI7SORMcoFdIbtJGBj681fZGhEvB8R6IdcdxjYWYWz4fmRJPFuRVGNk7y4rOH2LxnG2dPTJPGCcUoIiOnnyZxjBQGazRJPMD3ChgdI6QEp3KPfU+BcAz6Mf20i5TgtMMJhxd4xFqjjAMULi898C/x1Q3gnMGJPFAtEAJHXkhorVEShMztmJ0zWGPAgPIcSToAfKS2eFZiPYUxFmMNOHLKlpWARgiBUi9b+76UEdQ9zKolXu4jfI/RXZPUyooTjw84Zc8xvGGc1toc/agLRbhm6zUcWjlI69QCAXXe+rbb+bPf/wKbbyty5myf8ZH97Nwdcujbsyy2UsYuX0dSmaPeVAzXbqWj5jn2yHPs//AGZg5nrLUzQiXJvA5ZPM/xkzOMDNcYWWfoDPpkWnNybpZqMMz58xcY3z6FtR5ZLFhOEkRPkxpNMRpClEOalVEq5SYDFXNmeY1r917BVddeyeP3/gCvPMSurdv54n/7Pt3WMjrNqA8NM7llE5/962d41zuvZjA3y4nleUwvobZulHe87k4+/8WvMjU1icQhpCP1HYPAsXm8yMX5FZqNMlG1Stm0ePZomze+7W0cPXmKqeYIzl1yYTKWUqEI/wRF6x+cUJ1zhJ7KCwHlYZ3CWY24NGlUwsup+jJ/lzaaNUqNKo1ihaTTI1QegedR8gts2LSR6mpENkg5d/oC+w9cw5/+1SfpmDKX7b0DJWFsdD3nlyQ/9xu/T295ml/54Lup+Dfx4Q99BKnguqt2s3NqgmTmLIO0w81jYwzVStw+ei1KCVIrSLIMF1SYuXAUEYXoIGL7NddiXJ6Dp3UKSlKu1HJ735kXuGLnlUiR4DxodfpEUZlC4Ofvn0aBEyemqe/ZSLM+wgl7jsxmIPPvxvhok7uu3YtfLtCPEzYXBEttTZYZOms9CvUmy61l0qESf/+ZP0P4ikIU5fe1Wsf3Fb
```
Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: I would just show learned & empirical accuracy- and be clear the latter is just over the dev set!

File: tutorials/images/Images_Tutorial.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajratner) [reply tutorials/images/Images_Tutorial.ipynb]: Actually, since this is already outside of the Snorkel data structure, I would:
(1) Say explicitly that these are *probabilistic training labels* that can be used to train any model
(2) Say for example in TF, you'd use a cross-entropy loss (https://www.tensorflow.org/versions/r1.1/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits) to handle the fact that they are probabilistic

File: tutorials/images/Images_Tutorial.ipynb [hunk 11]
```diff
@@ -0,0 +1,758 @@
...
...60ctMjMGH7riVjVqPB+6+mz9ZPss7j99OY+sc1f4wB8oOMoqpNTaoJZb58YAoX2KOFi9farK9tMQdb93LsrFsX21w68EZSNpcXK6ya9801U3BQ/cdR+gcSI/t3hKOVUjRptst0uivouOQRt+n3W1QHt+PlEWE10KkAcpk2RyOymfNs7FICxoJaQ2Vy4EBpRyK5TKBG9CNYHjEZ2tjFCeXw+u0GCtOk3oxSSvBzaVc6sTMCUHf6aJfP0HS2WYoKCK9Cut6k6GVBs5EQDEt0KhDS6ZYLbGBh+c7uFKR2hSERNhBBpQVaBOjlERrlf0sLY4SCL+IjuNsmGMsURKhpI+UApNmFJDMjlIiRXYNizBgqOjwiD/BTz5XpzAyxX8cuZnbN06y9etj9HsjLKy+ynjZY2OrSrDQJ1xrssdK1rptWp7myL4ZNio1JsaHqBZyDL1tgvmmR61b59LVS7zvHW/j5fPnuGnfQf7TK08xVbS8triEVAnkhpm8B7jqMubt4oCx+LfuY2imzB+9voCxQ4zlYiKVMNrt8b/f+VZw8lgb4ftllAu1KOFUpcrNx/cj1rcIwxAThISthJr0scLHpH2aseH7PvJRXnr5HKsrJ3jy8Wf5R9//U/z2U6/wSrWOXuyTm18hLo/yxU+dZmmtjjv0CscO7uGGm/Yi4iYnTi4zvWuMxmYfXJ/5fSVyYY/b7xyjk7aoLPgc3RMz09nmi+dqVOQQe45OQmJoxwmLl3yCsSHOvLbGppAc3d1ldnyW5x59md6KYd/BHJ97+E/ptgWBNHSHpvnNP32Jkvs8yy3BP/ux+/+e7yjfqb9LmbRL3i8gHEWn3YWB9q6Qc+l2IgJPEgYhrieIe13KowV63ZhcQSKFg5JFHCdFJ5BGKTp1cYzFegIrXVAglHizoYwQBK6DFzhgE4SIcKRPGIZUt1YgCSkUivhK0Cv4NDuLFEqK9mrI669v0eu3cQgZnXQpqYh3vGcPQjd49fk+i4uWyX0t7j/QZqSwh6XzSxijKAYByyf7fNeH5xkausyX/rbG3vkxGpUCUsL89BTPPdek0VkgDF1C7xz53DCVldMc2ftOLp5b4cTJ83juFVznBkI3x4fe8094+vlnyIVgJEzOBVQaCtcBt1+m0Vyn1snT6Dto2WR0+m4urp+h6OUxrRb01rhp/80wNE7c6VMSHkZZRJSg949DFDO
```
Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: Edit:
```
You can then use these training labels to train any standard discriminative model, such as [a state of the art ResNet](POINTER HERE), which should learn to generalize beyond the LF's we've developed!

The only change needed from standard procedure is to deal with the fact that the training labels Snorkel generates are _probabilistic_ (i.e. for the binary case, in [0,1])- luckily, this is a one-liner in most modern ML frameworks! For example in TensorFlow, you can use the cross-entropy loss [(as defined here, in Tensorflow)](https://www.tensorflow.org/versions/r1.1/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits).
```

Comment (ajratner) [inline tutorials/images/Images_Tutorial.ipynb]: Edit:
```
Note that usually we have Snorkel manage our data using its ORM database backend, in which case we use the `LabelAnnotator` from the `snorkel.annotations` module (LINK TO TUTORIAL HERE). However here, we show how to construct labeling matrices manually- useful when managing your data outside of Snorkel, as is the case with our image data here!
```

**Comment by ajratner** [comment]
@vincentschen added two edits, once you put those in looks ready to merge- thanks!!
